### PR TITLE
DAOS-11829 object: add a DAOS API to allow users to set an anchor at …

### PR DESCRIPTION
**Pull request by mchaarawi**
…specific key

- add also DFS API on top of that to allow readdir to resume at an anchor set at a particular dir entry.

Required-githooks: true

Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by mchaarawi** [comment]
ping reviewers (@liuxuezhao @wangdi1 @jolivier23)

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

File: src/object/cli_obj.c
Comment (wangdi1) [inline src/object/cli_obj.c:6191]: This won't work for parity rotation EC objects,  since EC parity object actually allocate sub anchors for each ec shards within one group, and then track these sub anchors individually.  

So I assume you will need allocate sub anchors for all shards within one EC group, and set anchors individually on each shard.  see obj_list_common()->sub_anchors_prep().

Hmm, set anchor seems only useful for directory, but since EC object is rarely used for directory, probably just disable this for EC object?

Comment (mchaarawi) [reply src/object/cli_obj.c:6191]: on obj_list_common() we call sub_anchors_prep() only if shard_cnt > 1
here shard_cnt is always 1 right? maybe i don't understand..
but i cannot assume that users would not use EC for directory im afraid

Comment (wangdi1) [reply src/object/cli_obj.c:6191]: >on obj_list_common() we call sub_anchors_prep() only if shard_cnt > 1
here shard_cnt is always 1 right?

shard_cnt will be > 1 if parity rotation is enabled or all parity failed (in degraded mode).

> but i cannot assume that users would not use EC for directory im afraid

it can return failure? since this API is specifically for directory, which will rarely use EC object anyway. or I miss other use case?

Comment (mchaarawi) [reply src/object/cli_obj.c:6191]: we cannot assume that users will always use non-EC for directory.
let's say a middleware layer like dfuse uses this function to implement opendir/seekdir/readdir. then a user creates an EC directory (above dfuse control). normal readdir will not work anymore on this directory because it calls this function. so we shouldn't assume this at this layer.

Comment (wangdi1) [reply src/object/cli_obj.c:6191]: >let's say a middleware layer like dfuse uses this function to implement opendir/seekdir/readdir. then a user creates an EC directory (above dfuse control).

I understand the object type is not controlled by dfuse, though the directory object type can be controlled by hint/property?  so as long as we make it clear in our document, it should be fine? Besides, this does not seems a "general" API to me? But I might miss some use cases other than directory entry fetch?

If you insist to implement this for EC object, then it will need prepare the top anchor to create sub anchors for each shard as obj_list_common(),  then forward the request to all shards, in complete_callback(), it will set the those reply anchors in each sub_anchors individually see anchor_update_check_eof()->update_sub_anchor_cb().

So I am not sure if it is worth to implement these complexity for EC object, unless this api will be used to implement standard dfs_api() ?

Comment (jolivier23) [reply src/object/cli_obj.c:6191]: If the dkey tree exists on a shard, the da_buf will be the same independent of which shard fills it in.  So if there is any extra work to do there to fill out multiple subanchors, I think it can be done independently.   But I guess the issue may be that you don't know which shard may have the dkey tree allocated so that is why you need to send to all?

Comment (jolivier23) [reply src/object/cli_obj.c:6191]: In fact, it doesn't really even matter which shard you send it to if a tree exists for the dkey/akey you are trying to turn into an anchor.

Comment (jolivier23) [reply src/object/cli_obj.c:6191]: So is there no way to just reject EC object class for directories?  It is kind of terrible and probably not what the user actually wants.   It's really no different than replication, right?

Actually, that brings up an interesting point.   This should work as is if the EC object is just using replication (e.g. no full stripe writes), no?

Comment (mchaarawi) [reply src/object/cli_obj.c:6191]: ok so turns out with current master, we cannot event do dkey_list on EC object - get a segfault. so maybe that sub-anchor stuff is already broken? @wangdi1 ?

Comment (wangdi1) [reply src/object/cli_obj.c:6191]: hmm, we have some simple ec enumeration tests in daos_obj.c & daos_obj_ec.c. Maybe there is some bugs there. stack trace? create a ticket?

Comment (mchaarawi) [reply src/object/cli_obj.c:6191]: yes already done:
https://daosio.atlassian.net/browse/DAOS-12063

Comment (mchaarawi) [reply src/object/cli_obj.c:6197]: @wangdi1 after reviewing this more, in this case since we have a dkey passed in for the set_anchor, shard_cnt will always be 1. in this case, for enumeration, nothing is done for the sub anchor:
https://github.com/daos-stack/daos/blob/master/src/object/cli_obj.c#L6048
So im still confused by all this subanchor stuff and what it has to do with this case.. i do see that if we use EC directory, the enumeration does not start from the correct place though, so there is something wrong with this and EC.

Comment (mchaarawi) [reply src/object/cli_obj.c:6197]: OK i think i figured out why shard_nr was one in my case. it should be 4 but it was getting overwritten. now that i have this, i need to figure out how to allocate the subanchor, since in the case of enumeration, it relies on the user provided sgl, but in this case here, there is no sgl since the user is not doing enumeration at this point. any idea?

Comment (mchaarawi) [reply src/object/cli_obj.c:6197]: actually i was wrong. shard_cnt is always returned as 1 here:
https://github.com/daos-stack/daos/blob/master/src/object/cli_obj.c#L5860

Comment (mchaarawi) [reply src/object/cli_obj.c:6255]: ok i see. updated the code to support EC object. i pulled in your fix for enumeration for EC object to test it and it works fine with EC directory (tested locally). probably should land your PR https://github.com/daos-stack/daos/pull/10834 first then rebase this after.

**Review by jolivier23 — DISMISSED** [review]
Looks like it needs rebase now

**Review by jolivier23 — DISMISSED** [review]

File: src/object/cli_obj.c
Comment (wangdi1) [inline src/object/cli_obj.c:6255]: probably it should use the shard_cnt got from obj_ec_get_parity_or_alldata_shard(), in case parity rotation is not enabled (shard_cnt = 1), and also obj_get_grp_size() will also include the extending shard during online OSA, which is not needed for enumeration.

File: src/object/cli_obj.c
Comment (wangdi1) [inline src/object/cli_obj.c:4043]: Hmm, seems no different with other key enumeration operations.  I think obj_list_comp() might just work by using obj_list_dkey_cb().

**Review by wangdi1 — DISMISSED** [review]
Probably need some update, though it can be done once PR-10498 is landed, not sure which one will be landed first.

File: src/object/cli_obj.c
Comment (mchaarawi) [reply src/object/cli_obj.c:4043]: better to keep things separate for better readability?

File: src/object/cli_obj.c
Comment (mchaarawi) [reply src/object/cli_obj.c:6255]: well i think shard_cnt was always returned as 1 so i had to do this, but i will check again to make sure.

Comment (mchaarawi) [reply src/object/cli_obj.c:6297]: ok i confirmed that this is required as obj_ec_get_parity_or_alldata_shard always returning 1 for shard_cnt.. since i assume dkey is not NULL. so to be able to go query all shards, we have to do this.

Comment (wangdi1) [reply src/object/cli_obj.c:6297]: ah, hmm, probably should be

if (obj_ec_parity_rotate_enable())
         shard_cnt = daos_oclass_grp_size();

**Review by wangdi1 — APPROVED** [review]

**Comment by mchaarawi** [comment]
@wangdi1 only 1 failure here in: FTEST_osa.OSAOnlineExtend
that should be unrelated to this PR, right?

**Comment by wangdi1** [comment]
> @wangdi1 only 1 failure here in: FTEST_osa.OSAOnlineExtend
> that should be unrelated to this PR, right?

yes, DAOS-12031

**Comment by mchaarawi** [comment]
@jolivier23 need another review from you please since i had to rebase with conflicting PR

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**
